edit/candidate.go

```diff
@@ -101,39 +101,29 @@ func outputComplexCandidate(ec *eval.EvalCtx,
 	ec.OutputChan() <- c
 }
 
-func (ed *Editor) filterAndCookCandidates(ev *eval.Evaler, completer string, pattern string,
-	cands []rawCandidate, q parse.PrimaryType) ([]*candidate, error) {
+func (ed *Editor) filterAndCookCandidates(ev *eval.Evaler, matcher eval.CallableValue, pattern string,
+	cands chan eval.Value, q parse.PrimaryType) ([]*candidate, error) {
 
-	matcher, ok := ed.lookupMatcher(completer)
-	if !ok {
-		return nil, errMatcherMustBeFn
-	}
-
-	input := make(chan eval.Value, len(cands))
 	ports := []*eval.Port{
-		{Chan: input}, {File: os.Stdout}, {File: os.Stderr}}
+		{Chan: cands}, {File: os.Stdout}, {File: os.Stderr}}
 	ec := eval.NewTopEvalCtx(ev, "[editor matcher]", "", ports)
 
-	// TODO(xiaq): Revise the API of completers so that the complter write
-	// directly to a channel and run concurrently with the matcher.
-	for _, cand := range cands {
-		input <- eval.String(cand.text())
-	}
-	close(input)
-
 	args := []eval.Value{eval.String(pattern)}
 	values, err := ec.PCaptureOutput(matcher, args, eval.NoOpts)
 	if err != nil {
+		logger.Printf("tw; %s\n", err)
 		return nil, err
-	} else if len(values) != len(cands) {
-		return nil, errIncorrectNumOfResults
 	}
 
 	var filtered []*candidate
-	for i, value := range values {
-		if eval.ToBool(value) {
-			filtered = append(filtered, cands[i].cook(q))
+	for _, value := range values {
```

Comment:
The matcher needs to be aware of the actual types of the inputs, as you have done in the type switch inside `wrapMatcher`. Right now user-defined functions cannot access the `text()` method of `rawCandidate`, and this makes custom matcher in elv no longer possible.

We could introduce a way to access the `text()` result from elv, for instance by adding an `get-text-for-matching` builtin, and rely on matchers to use it, but this results in a less flexible API. Exposing only the text to filter to the matcher creates more work for the caller, but it also gives more control to the caller.

For instance, we do not only want user-customizable matchers for completion candidates, but also for items in listing modes (e.g. directory history item, command history item). To make matchers work for listing items, we can simply extend `get-text-for-matching` to work on listing items as well; for instance, calling it on a directory history item will simply give the directory path. But what if we add some more complex listing modes, where a listing item has multiple fields that can be filtered on -- e.g. a "process listing mode", where each process has fields like user, tty, command name? Then we have to make a choice of *which* of the fields `get-text-for-matching` returns, and by doing this you take away the ability to match on all other fields. Instead, if we only give strings to the matcher, the caller can always decide which field to give to the matcher. This is what I mean by saying that supplying strings gives more control to the caller.

I cannot think of a good use case where allowing matchers to add new candidates is useful; do you have one in mind? It also sounds like a generally bad idea to allow matchers to do too many things.

The original raw candidates come from the `rawCands` channel, and you can add another goroutine that feeds the `text()` of those raw candidates and collect the output.